# Stage 2. Compile summary information about the data
The purpose of this stage is to investigate and understand the data before starting EDA. Main tasks for this stage:
1. Importing dataset
2. Identifying relevant data structures and summarizing data
3. Extracting information from columns
4. Combining or modifying data structures to create meaningful variables


1. Importing dataset 

In [2]:
# import all necessary packages
import pandas as pd
import numpy as np

In [3]:
# loading dataset into a dataframe
df = pd.read_csv('waze_dataset.csv')

2. Identifying relevant data structures and summarizing data

For that, I use methods head() and info() to get general information about the dataframe structure, size, data types of content and the presence of missing values.

In [4]:
df.head(10)

ID     label  sessions  drives  total_sessions  n_days_after_onboarding  \
0   0  retained       283     226      296.748273                     2276   
1   1  retained       133     107      326.896596                     1225   
2   2  retained       114      95      135.522926                     2651   
3   3  retained        49      40       67.589221                       15   
4   4  retained        84      68      168.247020                     1562   
5   5  retained       113     103      279.544437                     2637   
6   6  retained         3       2      236.725314                      360   
7   7  retained        39      35      176.072845                     2999   
8   8  retained        57      46      183.532018                      424   
9   9   churned        84      68      244.802115                     2997   

   total_navigations_fav1  total_navigations_fav2  driven_km_drives  \
0                     208                       0       2628.845068   
1                      19                      64      13715.920550   
2                       0                       0       3059.148818   
3                     322                       7        913.591123   
4                     166                       5       3950.202008   
5                       0                       0        901.238699   
6                     185                      18       5249.172828   
7                       0                       0       7892.052468   
8                       0                      26       2651.709764   
9                      72                       0       6043.460295   

   duration_minutes_drives  activity_days  driving_days   device  
0              1985.775061             28            19  Android  
1              3160.472914             13            11   iPhone  
2              1610.735904             14             8  Android  
3               587.196542              7             3   iPhone  
4              1219.555924             27            18  Android  
5               439.101397             15            11   iPhone  
6               726.577205             28            23   iPhone  
7              2466.981741             22            20   iPhone  
8              1594.342984             25            20  Android  
9              2341.838528              7             3   iPhone

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14999 entries, 0 to 14998
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   ID                       14999 non-null  int64  
 1   label                    14299 non-null  object 
 2   sessions                 14999 non-null  int64  
 3   drives                   14999 non-null  int64  
 4   total_sessions           14999 non-null  float64
 5   n_days_after_onboarding  14999 non-null  int64  
 6   total_navigations_fav1   14999 non-null  int64  
 7   total_navigations_fav2   14999 non-null  int64  
 8   driven_km_drives         14999 non-null  float64
 9   duration_minutes_drives  14999 non-null  float64
 10  activity_days            14999 non-null  int64  
 11  driving_days             14999 non-null  int64  
 12  device                   14999 non-null  object 
dtypes: float64(3), int64(8), object(2)
memory usage: 1.5+ MB


As we see from the above results, the column 'label' includes missing values in 700 rows. Let's investigate them in detail. 

2.1. To understand if missing data is significant for further analysis, I compare the summary statistics of the 700 rows that have missing values in column 'label' with summary statistics of the rows that don't have any missing values.

In [23]:
# Rows with null values
label_empty=df[pd.isnull(df['label'])]

# Summary stats of rows with null values
label_empty.describe()

ID    sessions      drives  total_sessions  \
count    700.000000  700.000000  700.000000      700.000000   
mean    7405.584286   80.837143   67.798571      198.483348   
std     4306.900234   79.987440   65.271926      140.561715   
min       77.000000    0.000000    0.000000        5.582648   
25%     3744.500000   23.000000   20.000000       94.056340   
50%     7443.000000   56.000000   47.500000      177.255925   
75%    11007.000000  112.250000   94.000000      266.058022   
max    14993.000000  556.000000  445.000000     1076.879741   

       n_days_after_onboarding  total_navigations_fav1  \
count               700.000000              700.000000   
mean               1709.295714              118.717143   
std                1005.306562              156.308140   
min                  16.000000                0.000000   
25%                 869.000000                4.000000   
50%                1650.500000               62.500000   
75%                2508.750000              169.250000   
max                3498.000000             1096.000000   

       total_navigations_fav2  driven_km_drives  duration_minutes_drives  \
count              700.000000        700.000000               700.000000   
mean                30.371429       3935.967029              1795.123358   
std                 46.306984       2443.107121              1419.242246   
min                  0.000000        290.119811                66.588493   
25%                  0.000000       2119.344818               779.009271   
50%                 10.000000       3421.156721              1414.966279   
75%                 43.000000       5166.097373              2443.955404   
max                352.000000      15135.391280              9746.253023   

       activity_days  driving_days  
count     700.000000    700.000000  
mean       15.382857     12.125714  
std         8.772714      7.626373  
min         0.000000      0.000000  
25%         8.000000      6.000000  
50%        15.000000     12.000000  
75%        23.000000     18.000000  
max        31.000000     30.000000

In [24]:
# Rows without null values
label_full=df[pd.notnull(df['label'])]

# Summary stats of rows without null values
label_full.describe()

ID      sessions        drives  total_sessions  \
count  14299.000000  14299.000000  14299.000000    14299.000000   
mean    7503.573117     80.623820     67.255822      189.547409   
std     4331.207621     80.736502     65.947295      136.189764   
min        0.000000      0.000000      0.000000        0.220211   
25%     3749.500000     23.000000     20.000000       90.457733   
50%     7504.000000     56.000000     48.000000      158.718571   
75%    11257.500000    111.000000     93.000000      253.540450   
max    14998.000000    743.000000    596.000000     1216.154633   

       n_days_after_onboarding  total_navigations_fav1  \
count             14299.000000            14299.000000   
mean               1751.822505              121.747395   
std                1008.663834              147.713428   
min                   4.000000                0.000000   
25%                 878.500000               10.000000   
50%                1749.000000               71.000000   
75%                2627.500000              178.000000   
max                3500.000000             1236.000000   

       total_navigations_fav2  driven_km_drives  duration_minutes_drives  \
count            14299.000000      14299.000000             14299.000000   
mean                29.638296       4044.401535              1864.199794   
std                 45.350890       2504.977970              1448.005047   
min                  0.000000         60.441250                18.282082   
25%                  0.000000       2217.319909               840.181344   
50%                  9.000000       3496.545617              1479.394387   
75%                 43.000000       5299.972162              2466.928876   
max                415.000000      21183.401890             15851.727160   

       activity_days  driving_days  
count   14299.000000  14299.000000  
mean       15.544653     12.182530  
std         9.016088      7.833835  
min         0.000000      0.000000  
25%         8.000000      5.000000  
50%        16.000000     12.000000  
75%        23.000000     19.000000  
max        31.000000     30.000000

Comparing the above tables, there are no significant differences in the mean and standard deviation of the two datasets. It means that there won't be any significant impact on the result of the analysis if we include or do not include these data in further analysis

2.2. Next, I check the two datasets with respect to the 'device' variable.

In [25]:
# Rows with null values by device
label_empty['device'].value_counts()

device
iPhone     447
Android    253
Name: count, dtype: int64

In [26]:
# Percentage of iPhone and Android users in rows with null values
device_null=label_empty['device'].value_counts()/len(label_empty)
print(device_null)

device
iPhone     0.638571
Android    0.361429
Name: count, dtype: float64


In [27]:
# Percentage of iPhone and Android users in full dataset
device_full=df['device'].value_counts()/len(df)
print(device_full)

device
iPhone     0.644843
Android    0.355157
Name: count, dtype: float64


As we see, the percentage of missing values by each device is consistent with their representation in the dataset overall.

So I can suggest that there are no patterns in the appearance of the missing values and they appear randomly.

3. Once I have understood that missing values in the 'label' variable are random and not crucial for further analysis, I can analyze this variable more deeply.
First of all, I investigate and compare the number and percentages of users who churned vs. those who were retained

In [28]:
# Number of churned vs. retained users
label_counts=df['label'].value_counts()
print(label_counts)
print()
label_percentage=df['label'].value_counts(normalize=True)
print(label_percentage)

label
retained    11763
churned      2536
Name: count, dtype: int64

label
retained    0.822645
churned     0.177355
Name: proportion, dtype: float64


Next, I compare the medians of each variable for churned and retained users. The reason for calculating the median and not the mean is that I don't want to take outliers into account

In [29]:
# The medians for grouped data by `label`
group_label_median=df.groupby(['label']).median(numeric_only=True)
group_label_median

ID  sessions  drives  total_sessions  n_days_after_onboarding  \
label                                                                         
churned   7477.5      59.0    50.0      164.339042                   1321.0   
retained  7509.0      56.0    47.0      157.586756                   1843.0   

          total_navigations_fav1  total_navigations_fav2  driven_km_drives  \
label                                                                        
churned                     84.5                    11.0       3652.655666   
retained                    68.0                     9.0       3464.684614   

          duration_minutes_drives  activity_days  driving_days  
label                                                           
churned               1607.183785            8.0           6.0  
retained              1458.046141           17.0          14.0

As a result, I have got an interesting snapshot of the two groups of users, churned vs. retained:

Users who churned averaged ~3 more drives in the last month than retained users, but retained users used the app on over twice as many days as churned users in the same time period.

The median churned user drove ~200 more kilometers and 2.5 more hours during the last month than the median retained user.

It seems that churned users had more drives in fewer days, and their trips were farther and longer in duration. There is a probability that the big part of churned users are the truck drivers and maybe they have other requirements and expectations of the Waze app.

4. Next, I calculate additional variables that will help me to describe and understand the behavior of users better

In [13]:
# The median kilometers per drive in the last month for both retained and churned users
print('Median kilometers per drive:')
group_label_median['driven_km_drives']/group_label_median['drives']

Median kilometers per drive:


label
churned     73.053113
retained    73.716694
dtype: float64

In [14]:
# The median kilometers per driving day in the last month for both retained and churned users
print('Median kilometers per driving day:')
group_label_median['driven_km_drives']/group_label_median['driving_days']

Median kilometers per driving day:


label
churned     608.775944
retained    247.477472
dtype: float64

In [15]:
# The median number of drives per driving day in the last month for both retained and churned users
print('Median number of drives per driving day:')
group_label_median['drives']/group_label_median['driving_days']

Median number of drives per driving day:


label
churned     8.333333
retained    3.357143
dtype: float64

The median user who churned drove 608 kilometers each day they drove last month, which is almost 250% the per-drive-day distance of retained users. The median churned user had a similarly disproporionate number of drives per drive day compared to retained users.

Next, I examine whether there is an imbalance in how many users churned by device type.

In [21]:
# The number of Android and iPhone users in each group of users
df.groupby('label')['device'].value_counts()

label     device 
churned   iPhone     1645
          Android     891
retained  iPhone     7580
          Android    4183
Name: count, dtype: int64

In [22]:
# The percentage of Android and iPhone users in each group of users
df.groupby('label')['device'].value_counts(normalize=True)

label     device 
churned   iPhone     0.648659
          Android    0.351341
retained  iPhone     0.644393
          Android    0.355607
Name: proportion, dtype: float64

The ratio of iPhone users and Android users is consistent between the churned group and the retained group, and those ratios are both consistent with the ratio found in the overall dataset.

Key insights of Stage 2:
1. The dataset given by the client contains 12 variables that represent the main characteristics of 14999 unique users.
2. 700 users can't be classified as retained or churned because of the absence of information about their status in column 'label'. Regardless of that, there is no evidence that this missing information has non-random causes of appearance.  
3. The dataset contains 82% retained users and 18% churned users
4. There is a significant difference in the behavior of churned and retained users. 
 - Churned users averaged ~3 more drives in the last month than retained users.
 - Retained users used the app on over twice as many days as churned users in the last month.
 - The median churned user drove ~200 more kilometers and 2.5 more hours during the last month than the median retained user. 
 - The median user who churned drove 608 kilometers each day they drove last month, which is almost 250% the per-drive-day distance of retained users.

I can suggest that there are a lot of truckers among the churned users. As they are different from usual commuters, Waze should investigate their requirements and expectations of the Waze app in more detail. It could help to reduce the churn of users and increase the satisfaction rate of the Waze app.  
5. It is clear from the analysis that, regardless of whether a user churned or not, the users represented in this dataset are serious drivers. It would probably be safe to assume that this data does not represent typical drivers or commuters in general. According to the fact how much these users drive, I would recommend to to gather more data on these super-drivers. It's possible that the reason for their driving so much is also the reason why the Waze app does not meet their specific set of needs, which may differ from the needs of a more typical driver, such as a commuter. 